### ISSUE TITLE
Make the `uri` property of `RedirectException` nullable

### ISSUE THREAD

**Issue by brianquinlan** [dart-lang/sdk#49045]
# Change

Change:

```dart

class RedirectException implements HttpException {
-  Uri get uri => ...
+  Uri? get uri => ...
}
```

#  Rationale

This will allow us to cleanly fix https://github.com/dart-lang/sdk/issues/49012, which causes a `StateError` when there is no available `uri`.

# Impact

This will only break code that uses the `RedirectException` class directly and does not handle the `null` `uri` case.

Code that works with the `HttpException` base class will not need to be changed because `uri` is already nullable in that class.

This change does not break any Google code or any third-party code used by Google.

# Mitigation

Users must handle a possible `null` return from `RedirectException.uri`.

**Comment by a-siva** [comment]
//cc @lrhn

**Comment by itsjustkevin** [comment]
@vsmenon @Hixie @grouma Could you all take a look at this breaking change proposal?

**Comment by grouma** [comment]
I didn't find any usages of `RedirectException.uri` inside Google3. I assume this doesn't impact `HttpException` which has the non nullable `requestUrl` member. This is referenced in a handful of places.

**Comment by brianquinlan** [comment]
@grouma If you are referring to the `HttpException` in `...xbid/ui/common/exceptions/lib/exceptions.dart` then that class is in a completely different class hierarchy.

**Comment by Hixie** [comment]
I have no opinion on our change.

**Comment by brianquinlan** [comment]
Fixed in e5af733398d3d915dd794f765268c33428dd5729
